20220831-add-type-restrictions-to-json-syntax.md

```diff
@@ -0,0 +1,570 @@
+- [ListObjects Endpoint](https://openfga.dev/api/service#/Relationship%20Queries/ListObjects)
+- [What is a user?](https://openfga.dev/docs/concepts#what-is-a-user)
+- Floating User ID: A user identifier without a type (e.g. `anne`, `4` or `4179af14-f0c0-4930-88fd-5570c7bf6f59`)
+- Reverse Expansion: Normally [expand](https://openfga.dev/docs/interacting/relationship-queries#expand) takes an object and relation and returns the first leaves of users who are related. Reverse expand would be the opposite, taking a user and a relation and returning the objects which are related.
+- Representing public access: [`*`](https://openfga.dev/docs/concepts#how-do-i-represent-everyone)
+
+## Motivation
+
+### Why should we do this?
+
+#### Optimize ListObjects
+
+In order to implement some optimizations to the ListObjects endpoint mentioned [ListObjects RFC](https://github.com/openfga/rfcs/blob/main/20220714-listObjects-api.md), we should be able to traverse the relationship graph in reverse. We not only need to understand how objects are related to other objects, but also what type of objects can be related.
+
+#### Improve the Developer Experience
+
+At the moment, our authorization model does not allow users to indicate what the type they expect on a relation to be, leaving it ambiguous and potentially a cause of errors, especially when someone is reading the authorization model later without much context and where the original authors are not available. It is hard for someone to grok from first glance that a repository owner should be an organization or a user. Adding type restrictions allows developers to have an easier time in interpreting the model.
+
+### What is the expected outcome?
+
+- Using the extra type information we get from this change, we will be able to more easily traverse the graph in reverse, thus allowing us to provide a solution for reverse expansion and optimizing the ListObjects endpoint.
+- Users reading an authorization model will be able to better interpret its meaning (for example by understanding that repository owners must be either users or group members).
+
+## What it is?
+
+### Adding Type Restrictions
+
+This RFC proposes an update to the OpenFGA JSON syntax requiring users to indicate all the types of users that could be **directly related** to an object of a certain type through a particular relation.
+
+For example, our current syntax allows expressing the following:
+
+- documents have parents
+- repositories have owners
+
+But cannot express restrictions such as:
+
+- parents of a document have to be objects of type folder
+- owners of a repository have to be of type user or a userset of group members
+
+We introduce a `metadata` entry in the type definition would look like:
+
+```typescript
+type Metadata = {
+  relations?: Record<string, RelationMetadata>;
+};
+
+type RelationMetadata = {
+  allow_public?: Boolean;
+  directly_related_user_types?: DirectlyRelatedUserType[];
```

Comment:
I may suggest just a slight rename here.. `assignable_object_relations` or `assignable_relationship_refs` or `assignable_relationship_types`

Reply:
Assignable is nice, but we don't use it anywhere else as a term - so it is weird to have it in the spec.

I struggle a bit with the naming, because in its essence, this is neither assignable_object_relations nor directly_related_user_types.